module-system/sov-modules-macros/src/cli_parser.rs

```diff
@@ -25,49 +29,111 @@ impl CliParserMacro {
-                let field_name = field.ident.clone();
-                let field_name_string = field_name.to_string();
-                let encode_function_name = format_ident!("encode_{}_call", field_name_string);
-
-                // TODO:
-                // For the initial version, before complicating the macro,
-                // we're assuming that each module type in Runtime only has
-                // one generic. we're removing that and appending the concrete
-                // that's passed in from the macro.
-                // We need to fix this so that:
-                // 1. Determine which generic has the Context bound
-                // 2. Identify only that generic from the module type and replace it
-                // 3. Retain other generics
-
-                // Extract the type name
-                let type_path = match &field.ty {
-                    Type::Path(type_path) => {
-                        let mut segments = type_path.path.segments.clone();
-                        let last = segments.last_mut().expect("Impossible happened! A type path has no segments");
-                        last.arguments = PathArguments::None;
-                        Path { segments, ..type_path.path.clone() }
-                    },
-                    _ => return Err(syn::Error::new_spanned(field.ident, "expected a type path")),
-                };
-
-                let type_name_string = type_path.segments.last().unwrap().ident.to_string();
-
-                Ok(quote! {
-                #type_name_string => Ok({
-                    #ident::<#context_type>::#encode_function_name(
-                        serde_json::from_str::<<#type_path<#context_type> as sov_modules_api::Module>::CallMessage>(&call_data)?
-                    )
-                }),
-            })
-            })
-            .collect::<Result<Vec<_>, _>>()?;
+        let mut command_types = vec![];
+        let mut module_commands = vec![];
+        let mut module_args = vec![];
+        let mut match_arms = vec![];
+        let mut command_idents = vec![];
+        let mut parse_match_arms = vec![];
+
+        // Loop over the fields
+        for field in &fields {
+            if skip_fields.contains(&field.ident.to_string()) {
+                continue;
+            }
+            if let syn::Type::Path(type_path) = &field.ty {
+                let mut module_path = type_path.path.clone();
+                if let Some(segment) = module_path.segments.last_mut() {
+                    if let syn::PathArguments::AngleBracketed(angle_bracketed_data) =
+                        &mut segment.arguments
+                    {
+                        let mut new_args = Vec::new();
+                        for gen_arg in &angle_bracketed_data.args {
+                            match gen_arg {
+                                syn::GenericArgument::Type(syn::Type::Path(type_path))
+                                    if type_path.path.is_ident("C") =>
```

Comment:
We can't hardcode `C` here. Everything should work even when someone writes `<K: Context>` and we need unit tests for this.

Reply:
Fixed by the new macro